amazon/ion/sliceable_buffer.py

```diff
@@ -0,0 +1,164 @@
+            self._offset,
+            self.size + pair.length)
+
+    def read_byte(self):
+        """
+        Read the next byte from the buffer, return (byte, new buffer).
+
+        Raise IncompleteReadError if the buffer is empty.
+        """
+        size = self.size
+        chunks = self._chunks
+        offset = self._offset
+
+        try:
+            # assume that we have data, and that chunks are non-empty
+            (chunk, length) = chunks[0]
+        except IndexError:
+            raise IncompleteReadError("Buffer is empty!")
+
+        if length == offset + 1:
+            return chunk[offset], SliceableBuffer(chunks[1:], 0, size - 1)
+        else:
+            return chunk[offset], SliceableBuffer(chunks, offset + 1, size - 1),
+
+    def read_slice(self, n):
+        """
+        Read a slice of the buffer, return (slice, new buffer).
+
+        Raise IncompleteReadError if the slice cannot be fully read.
+
+        Chunks which are no longer readable are dropped from the new buffer.
+        Bytes are only copied if the read requires bridging chunks.
+        """
+        size = self.size
+        chunks = self._chunks
+        offset = self._offset
+        endpos = offset + n
+
+        if size < n:
+            raise IncompleteReadError(f'Buffer has size {size}, but {n} bytes were requested!')
+
+        # short-circuit when we can serve full read from first chunk
+        # optimizes for common case and simplifies accumulation loop
+        (chunk, length) = chunks[0]
+        if endpos < length:
+            return chunk[offset:endpos], SliceableBuffer(chunks, offset + n, size - n)
+        elif endpos == length:
+            return chunk[offset:], SliceableBuffer(chunks[1:], 0, size - n)
+
+        slices = [_ChunkPair(chunk[offset:], length - offset)]
+
+        # remaining and i are used to init the new buffer after the loop
+        remaining = endpos - length
+        i = 1
+        for (i, pair) in enumerate(chunks[1:], start=1):
+            (chunk, length) = pair
+            if remaining < length:
+                slices.append(_ChunkPair(chunk[:remaining], remaining))
+                break
+
+            slices.append(pair)
+            remaining -= length
+            if remaining == 0:
+                # move i forward to drop the chunk in the new buffer
+                i += 1
+                break
+
+        combined = bytearray(n)
+        cursor = 0
+        for (chunk, length) in slices:
+            combined[cursor:cursor + length] = chunk
+            cursor += length
+
+        return memoryview(combined), SliceableBuffer(chunks[i:], remaining, size - n)
+
+    def skip(self, n):
+        """
+        Skip max(n, size) bytes, return (skipped, new buffer).
+
+        Chunks which are no longer readable are dropped from the new buffer.
+        """
+        size = self.size
+        chunks = self._chunks
+        offset = self._offset
+        endpos = offset + n
+
+        if size <= n:
```

Comment:
Should `size < n` be `IncompleteReadError`?

Reply:
I agree with your insight. However, allowing the partial skip keeps memory usage lower when skipping a large container. That may be a premature optimization, but it is how the existing skip works and I didn't want to introduce a potential regression. Let me add a comment there.

Reply:
I agree that partial skip should be allowed, but should it skip what it can and then throw `IncompleteReadError`?

Reply:
Not really: if there are any changes to state they need to be returned from the call with a new buffer and not mutate the old one.